tools/scripts/bumpenvs.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
+"""
+Update all the environment image tags in the repository based on bumpenvs.yaml.
+
+Usage: bumpenvs.py path/to/bumpenvs.yaml
+"""
+
+import os
+import subprocess
+import sys
+from typing import List
+
+import yaml
+
+
+def samefile(a: str, b: str) -> bool:
+    return os.path.samefile(os.path.expanduser(a), os.path.expanduser(b))
+
+
+def list_git_paths_with_tag(tag: str, ignore_files: List[str]) -> List[str]:
+    cmd = ["git", "rev-parse", "--show-toplevel"]
+    r = subprocess.run(cmd, stdout=subprocess.PIPE, universal_newlines=True, check=True)
+
+    git_root = r.stdout.strip("\n")
+
+    cmd = ["git", "grep", "-z", "-l", tag, git_root]
+    r = subprocess.run(cmd, stdout=subprocess.PIPE, universal_newlines=True, check=True)
+    paths = r.stdout.strip("\0\n").split("\0")
+
+    return [p for p in paths if not any(samefile(p, ignore) for ignore in ignore_files)]
+
+
+def replace_tags(path: str, old_tag: str, new_tag: str) -> None:
+    with open(path) as f:
+        lines_in = list(f)
+
+    lines_out = [line.replace(old_tag, new_tag) for line in lines_in]
+
+    # Print git-style diffs.
+    for before, after in zip(lines_in, lines_out):
+        if before != after:
+            print(f"\x1b[31m-{before.rstrip()}\x1b[m", file=sys.stderr)
+            print(f"\x1b[32m+{after.rstrip()}\x1b[m", file=sys.stderr)
+            print(file=sys.stderr)
+
+    with open(path, "w") as f:
+        f.write("".join(lines_out))
```

Comment:
non-blocking: Using `f.writelines` here and `f.readlines` above may help with readability/explicitness.

Reply:
done.